pkg/proc/types.go

```diff
@@ -129,62 +131,128 @@ func (bi *BinaryInfo) loadPackageMap() error {
+			cu.Name, _ = entry.Val(dwarf.AttrName).(string)
+			compdir, _ := entry.Val(dwarf.AttrCompDir).(string)
+			if compdir != "" {
+				cu.Name = filepath.Join(compdir, cu.Name)
+			}
+			if ranges, _ := bi.dwarf.Ranges(entry); len(ranges) == 1 {
+				cu.LowPC = ranges[0][0]
+				cu.HighPC = ranges[0][1]
+			}
+			lineInfoOffset, _ := entry.Val(dwarf.AttrStmtList).(int64)
+			if lineInfoOffset >= 0 && lineInfoOffset < int64(len(debugLineBytes)) {
+				cu.lineInfo = line.Parse(compdir, bytes.NewBuffer(debugLineBytes[lineInfoOffset:]))
+			}
+			bi.compileUnits = append(bi.compileUnits, cu)
+
 		case dwarf.TagArrayType, dwarf.TagBaseType, dwarf.TagClassType, dwarf.TagStructType, dwarf.TagUnionType, dwarf.TagConstType, dwarf.TagVolatileType, dwarf.TagRestrictType, dwarf.TagEnumerationType, dwarf.TagPointerType, dwarf.TagSubroutineType, dwarf.TagTypedef, dwarf.TagUnspecifiedType:
 			if name, ok := entry.Val(dwarf.AttrName).(string); ok {
+				if !cu.isgo {
+					name = "C." + name
+				}
 				if _, exists := bi.types[name]; !exists {
 					bi.types[name] = entry.Offset
 				}
 			}
 			reader.SkipChildren()
+
 		case dwarf.TagVariable:
 			if n, ok := entry.Val(dwarf.AttrName).(string); ok {
+				if !cu.isgo {
+					n = "C." + n
+				}
 				bi.packageVars[n] = entry.Offset
 			}
+
 		case dwarf.TagSubprogram:
-			lowpc, ok1 := entry.Val(dwarf.AttrLowpc).(uint64)
-			highpc, ok2 := entry.Val(dwarf.AttrHighpc).(uint64)
+			ok1 := false
+			var lowpc, highpc uint64
+			if ranges, _ := bi.dwarf.Ranges(entry); len(ranges) == 1 {
+				ok1 = true
+				lowpc = ranges[0][0]
+				highpc = ranges[0][1]
+			}
+			name, ok2 := entry.Val(dwarf.AttrName).(string)
 			if ok1 && ok2 {
-				bi.functions = append(bi.functions, functionDebugInfo{lowpc, highpc, entry.Offset})
+				if !cu.isgo {
+					name = "C." + name
+				}
+				bi.Functions = append(bi.Functions, Function{
+					Name:  name,
+					Entry: lowpc, End: highpc,
+					offset: entry.Offset,
+					cu:     cu,
+				})
 			}
 			reader.SkipChildren()
+
 		}
 	}
-	sort.Sort(sortFunctionsDebugInfoByLowpc(bi.functions))
+	sort.Sort(compileUnitsByLowpc(bi.compileUnits))
+	sort.Sort(functionsDebugInfoByEntry(bi.Functions))
+
+	bi.LookupFunc = make(map[string]*Function)
+	for i := range bi.Functions {
+		bi.LookupFunc[bi.Functions[i].Name] = &bi.Functions[i]
+	}
+
+	bi.Sources = []string{}
+	for _, cu := range bi.compileUnits {
+		if cu.lineInfo != nil {
+			for _, fileEntry := range cu.lineInfo.FileNames {
+				bi.Sources = append(bi.Sources, fileEntry.Path)
+			}
+		}
+	}
+	sort.Strings(bi.Sources)
+	bi.Sources = uniq(bi.Sources)
 }
 
-func (bi *BinaryInfo) findFunctionDebugInfo(pc uint64) (dwarf.Offset, error) {
-	i := sort.Search(len(bi.functions), func(i int) bool {
-		fn := bi.functions[i]
-		return pc <= fn.lowpc || (fn.lowpc <= pc && pc < fn.highpc)
-	})
-	if i != len(bi.functions) {
-		fn := bi.functions[i]
-		if fn.lowpc <= pc && pc < fn.highpc {
-			return fn.offset, nil
+func uniq(s []string) []string {
```

Comment:
Kind of hard for me to follow this algorithm here. Can you explain a bit what's going on? It seems you're trying to return a unique list of strings? Why not use a map`[string]struct{}` as a set instead?

Reply:
It's the same as the unix command `uniq`:

```
       Filter  adjacent matching lines from INPUT (or standard input), writing
       to OUTPUT (or standard output).
```

so, yes, it's the same as using a set, only it doesn't allocate anything. I did it this way because I always do it this way, I can change it if you want.

Reply:
This will only filter successive duplicates, is that going to be an issue?

Reply:
It's called after a sort so all duplicates are consecutive.

Reply:
SGTM